geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/RangeHistogram.scala

```diff
@@ -91,43 +136,91 @@ class RangeHistogram[T](val attribute: Int, initialBins: Int, initialEndpoints:
 
   override def isEmpty: Boolean = bins.counts.forall(_ == 0)
 
   override def clear(): Unit = bins.clear()
 
-  override def equals(other: Any): Boolean = other match {
-    case that: RangeHistogram[T] => attribute == that.attribute && bins == that.bins // bins compares endpoints and length
+  override def isEquivalent(other: Stat): Boolean = other match {
+    case that: RangeHistogram[T] =>
+      attribute == that.attribute && bounds == that.bounds &&
+          java.util.Arrays.equals(bins.counts, that.bins.counts)
     case _ => false
   }
-
-  override def hashCode(): Int = Seq(attribute, bins).map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
 }
 
 object RangeHistogram {
 
+  def buffer[T](value: T): (T, T) = {
+    val buf = value match {
+      case v: String => (v, v + "z")
+      case v: Int    => (v - 1, v + 1)
+      case v: Long   => (v - 1, v + 1)
+      case v: Float  => (v - 1, v + 1)
+      case v: Double => (v - 1, v + 1)
+      case v: Date   => (v, new Date(v.getTime + 60000))
+      case v: Geometry =>
+        val env = v.getCentroid.buffer(1.0).getEnvelopeInternal
+        val min = GeometryUtils.geoFactory.createPoint(new Coordinate(env.getMinX, env.getMinY))
+        val max = GeometryUtils.geoFactory.createPoint(new Coordinate(env.getMaxX, env.getMaxY))
+        (min, max)
+    }
+    buf.asInstanceOf[(T, T)]
+  }
+  /**
+    * Creates a new binned array that encompasses the new value.
+    *
+    * Assumes that the value is not already within the bounds for the existing binned array.
+    */
+  def expandBins[T](value: T, old: BinnedArray[T])(implicit defaults: MinMax.MinMaxDefaults[T],ct: ClassTag[T]): BinnedArray[T] = {
+    val min = defaults.min(value, old.bounds._1)
+    val max = defaults.max(value, old.bounds._2)
+    val bins = BinnedArray[T](old.length, (min, max))
+    copyInto(bins, old)
+    bins
+  }
+
+  /**
+    * Gets new endpoints that encompass both arrays. If either array has empty values to start or end, the
+    * bounds will be trimmed down to be the start/end of non-empty values.
+    */
+  def checkEndpoints[T](left: BinnedArray[T], right: BinnedArray[T])(implicit defaults: MinMax.MinMaxDefaults[T]): (T, T) = {
+    val (lMin, lMax) = getActualBounds(left)
+    val (rMin, rMax) = getActualBounds(right)
+    (defaults.min(lMin, rMin), defaults.max(lMax, rMax))
+  }
+
   /**
-    * Creates expanded endpoints that cover both input endpoints
+    * Gets the bounds of the array that actually contain values.
     */
-  def expandEndpoints[T](one: (T, T), two: (T, T))(implicit defaults: MinMax.MinMaxDefaults[T]): (T, T) = {
-    val leftExpand = defaults.minmax(two._1, one._1, one._2)
-    defaults.minmax(two._2, leftExpand._1, leftExpand._2)
+  private def getActualBounds[T](bins: BinnedArray[T])(implicit defaults: MinMax.MinMaxDefaults[T]): (T, T) = {
+    val minIndex = bins.counts.indexWhere(_ != 0)
+    val maxIndex = bins.counts.length - bins.counts.reverse.indexWhere(_ != 0) - 1
```

Comment:
If there are 10 bin counts, all of which are empty, then ```maxIndex = 10 - -1 - 1 = 10```, which should cause an overflow on the line computing ```max```, shouldn't it?